# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/remotebrowserutils/RemoteWebNavigation.jsm
# Commit: 04d9dba72f22
# Full Hash: 04d9dba72f221b5808f077e3aac5522147008438
# Author: Jim Porter <jporter@mozilla.com>
# Date: 2019-06-04 21:42:40
# Regressor Bug: 1552017
# File Overlap Count: 1
# Description:
#   Bug 1552017 - Expand the kinds of URLs that can cancel content JS when navigating; r=smaug
#   
#   This patch makes several changes to the kinds of URLs where we can cancel
#   content JS when navigating between them:
#   
# ==============================================================================

diff -r 25b1ee5ee458 -r 04d9dba72f22 toolkit/components/remotebrowserutils/RemoteWebNavigation.jsm
--- a/toolkit/components/remotebrowserutils/RemoteWebNavigation.jsm	Tue Jun 04 16:44:29 2019 +0000
+++ b/toolkit/components/remotebrowserutils/RemoteWebNavigation.jsm	Tue Jun 04 16:19:27 2019 +0000
@@ -12,10 +12,6 @@
 ChromeUtils.defineModuleGetter(this, "E10SUtils",
   "resource://gre/modules/E10SUtils.jsm");
 
-function makeURI(url) {
-  return Services.io.newURI(url);
-}
-
 function RemoteWebNavigation() {
   this.wrappedJSObject = this;
   this._cancelContentJSEpoch = 1;
@@ -77,14 +73,18 @@
   },
   loadURI(aURI, aLoadURIOptions) {
     let uri;
+    try {
+      let fixup = Cc["@mozilla.org/docshell/urifixup;1"].getService();
+      let fixupFlags = fixup.webNavigationFlagsToFixupFlags(
+        aURI, aLoadURIOptions.loadFlags);
+      uri = fixup.createFixupURI(aURI, fixupFlags);
 
-    // We know the url is going to be loaded, let's start requesting network
-    // connection before the content process asks.
-    // Note that we might have already setup the speculative connection in some
-    // cases, especially when the url is from location bar or its popup menu.
-    if (aURI.startsWith("http:") || aURI.startsWith("https:")) {
-      try {
-        uri = makeURI(aURI);
+      // We know the url is going to be loaded, let's start requesting network
+      // connection before the content process asks.
+      // Note that we might have already setup the speculative connection in
+      // some cases, especially when the url is from location bar or its popup
+      // menu.
+      if (uri.schemeIs("http") || uri.schemeIs("https")) {
         let principal = aLoadURIOptions.triggeringPrincipal;
         // We usually have a triggeringPrincipal assigned, but in case we
         // don't have one or if it's a SystemPrincipal, let's create it with OA
@@ -97,10 +97,10 @@
           principal = Services.scriptSecurityManager.createCodebasePrincipal(uri, attrs);
         }
         Services.io.speculativeConnect(uri, principal, null);
-      } catch (ex) {
-        // Can't setup speculative connection for this uri string for some
-        // reason (such as failing to parse the URI), just ignore it.
       }
+    } catch (ex) {
+      // Can't setup speculative connection for this uri string for some
+      // reason (such as failing to parse the URI), just ignore it.
     }
 
     let cancelContentJSEpoch = this._cancelContentJSEpoch++;
@@ -138,7 +138,7 @@
   _currentURI: null,
   get currentURI() {
     if (!this._currentURI) {
-      this._currentURI = makeURI("about:blank");
+      this._currentURI = Services.io.newURI("about:blank");
     }
 
     return this._currentURI;
